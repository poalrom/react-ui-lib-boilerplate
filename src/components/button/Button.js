// @flow

import * as React from 'react';
import classnames from 'classnames';
import omit from 'object.omit';
import './Button.styl';

type Props = {
  /** Тип кнопки */
  type?: 'button' | 'submit',
  /** Функция, вызываемая по клику на кнопку, если она не выключена */
  onClick?: () => void,
  /** Флаг лоадера */
  loading?: boolean,
  /** Дополнительные классы */
  className?: string,
  /** Элементы, вкладываемые в кнопку */
  children?: React.Node,
  /** Флаг выключения */
  disabled?: boolean,
};

const Button = (props: Props) => {
  const buttonProps = omit(props, ['className', 'loading']);
  return (
    <button
      className={classnames(
        'button',
        props.className,
        {
          button_loading: props.loading,
          button_disabled: props.disabled,
        },
      )}
      {...buttonProps}
    >
      {props.children}
    </button>
  );
};

Button.defaultProps = {
  type: 'button',
  loading: false,
  className: '',
  children: null,
  disabled: false,
  onClick: () => {},
};

export default Button;
