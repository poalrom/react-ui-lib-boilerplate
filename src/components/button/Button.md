### Кнопка, самая обыкновенная

Можно кликнуть, навести или перейти с помощью `Tab`
```js
<Button onClick={() => {alert('Клик на кнопку')}}>Кликни на меня</Button>
```

Лоадер включается с помощью флага loading
```js
<Button loading={true} onClick={() => {alert('Подожди окончания загрузки')}}>Загружаю данные</Button>
```

Кнопка выключается с помощью флага disabled. При этом функция onClick не срабатывает
```js
<Button disabled={true} onClick={() => {alert('Никогда такого не было, и вот опять')}}>На меня не кликнуть =)</Button>
```