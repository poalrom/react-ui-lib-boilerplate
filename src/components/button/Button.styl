@keyframes spin
  100%
    -webkit-transform: rotate(360deg)
    transform: rotate(360deg)

.button
  background-color: hsl(221, 44%, 52%)
  border: none
  border-radius: 3px
  color: #fff
  cursor: pointer
  font-size: 15px
  line-height: 1.125
  padding: 12px 32px
  position: relative

  &:hover,
  &:focus
    background-color: hsl(221, 44%, 47%)
    outline: none

  &:active
    background-color: hsl(221, 44%, 42%)

  &_disabled,
  &_disabled:hover,
  &_disabled:focus,
  &_disabled:active
    background-color: hsl(221, 44%, 62%)
    cursor: auto

  &_loading
    &:after
      animation: spin 1s linear infinite
      background-image: url(loader.svg)
      content: ''
      display: block
      height: 16px
      position: absolute
      right: 8px
      top: 12px
      width: 16px
