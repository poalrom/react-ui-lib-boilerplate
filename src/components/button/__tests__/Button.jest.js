import React from 'react';
import { shallow } from 'enzyme';
import Button from '../Button';

test('render a button', () => {
  const wrapper = shallow(<Button>Simple button</Button>);
  expect(wrapper).toMatchSnapshot();
});

test('render a loading button', () => {
  const wrapper = shallow(<Button loading>Button with loader</Button>);
  expect(wrapper).toMatchSnapshot();
});

test('render a submit button', () => {
  const wrapper = shallow(<Button type="submit">Submit button</Button>);
  expect(wrapper).toMatchSnapshot();
});

test('render a button with custom classname', () => {
  const wrapper = shallow(<Button className="custom-button">Custom button</Button>);
  expect(wrapper).toMatchSnapshot();
});

test('render a disabled button', () => {
  const wrapper = shallow(<Button disabled>Disabled button</Button>);
  expect(wrapper).toMatchSnapshot();
});

test('onClick triggered', () => {
  const onClick = jest.fn();
  const wrapper = shallow(<Button onClick={onClick}>Clickable button</Button>);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  expect(onClick).toBeCalled();
});

test('onClick not triggered on disabled button', () => {
  const onClick = jest.fn();
  const wrapper = shallow(<Button disabled>Disabled button</Button>);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  expect(onClick).not.toBeCalled();
});
