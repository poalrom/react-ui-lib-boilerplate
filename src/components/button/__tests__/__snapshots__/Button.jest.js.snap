// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onClick not triggered on disabled button 1`] = `
<button
  className="button button_disabled"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Disabled button
</button>
`;

exports[`onClick triggered 1`] = `
<button
  className="button"
  disabled={false}
  onClick={[MockFunction]}
  type="button"
>
  Clickable button
</button>
`;

exports[`render a button 1`] = `
<button
  className="button"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Simple button
</button>
`;

exports[`render a button with custom classname 1`] = `
<button
  className="button custom-button"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Custom button
</button>
`;

exports[`render a disabled button 1`] = `
<button
  className="button button_disabled"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Disabled button
</button>
`;

exports[`render a loading button 1`] = `
<button
  className="button button_loading"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Button with loader
</button>
`;

exports[`render a submit button 1`] = `
<button
  className="button"
  disabled={false}
  onClick={[Function]}
  type="submit"
>
  Submit button
</button>
`;
